### PR TITLE
Enable custom pod specs for infra validator kubernetes serving pod

### PR DESCRIPTION
**Pull request by ConverJens**
This PR adds logic for enabling custom pod specs for infra validator kubernetes serving pod. It solves https://github.com/tensorflow/tfx/issues/3257.

### CODE REVIEW

**Comment by ConverJens** [comment]
The failing lints pre-dates this PR.

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py [hunk 1]
```diff
@@ -112,6 +112,10 @@ def __init__(
     self._label_dict = {
         _APP_KEY: _MODEL_SERVER_APP_LABEL,
     }
+    self._annotations_dict = {k_v_pair.k: k_v_pair.v for k_v_pair in
```
Comment (chongkong) [inline tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:115]: nit: I think this is a redundant copy, as we can directly use `overrides.annotations` and `overrides.environment_variables` as a `Mapping[str, str]` type.

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py [hunk 2]
```diff
@@ -208,6 +212,9 @@ def _BuildPodManifest(self) -> k8s_client.V1Pod:
     if isinstance(self._serving_binary, serving_bins.TensorFlowServing):
       env_vars_dict = self._serving_binary.MakeEnvVars(
           model_path=self._model_path)
+      for key, value in self._custom_env_vars_dict.items():
```
Comment (chongkong) [inline tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:215]: nit: how about simply

```python
if self._config.serving_pod_spec_overrides:
  env_vars_dict.update(
      self._config_serving_pod_spec_overrides.environment_variables)
```

**Review by chongkong — CHANGES_REQUESTED** [review]

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py (hunk 2 above)
Comment (ConverJens) [reply tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:215]: The reason for the current logic was that I didn't want the overrides to override the IV specific env vars such as tf servings MODEL.

I'm happy to change to your proposal if you think it's fine.

File: tfx/proto/infra_validator.proto [hunk 3]
```diff
@@ -119,6 +119,25 @@ message KubernetesConfig {
   //
   // Should be a positive integer. Default to 24 hours (86400).
   int32 active_deadline_seconds = 2;
+
+  // Optional.
+  // Additional Kubernetes pod specifications to apply to the serving pod
+  ServingPodSpecsOverrides serving_pod_specs_overrides = 3;
+}
+
+message ServingPodSpecsOverrides {
```
[the remark this answers is not in the document]
Comment (ConverJens) [reply tfx/proto/infra_validator.proto:128]: Agreed

File: tfx/proto/infra_validator.proto [hunk 4]
```diff
@@ -119,6 +119,25 @@ message KubernetesConfig {
...
+  // Optional.
+  // Kubernetes annotations to apply to the pod
+  repeated KeyValuePair annotations = 1;
```
[the remark this answers is not in the document]
Comment (ConverJens) [reply tfx/proto/infra_validator.proto:131]: Didn't now it existed but seems as a better choice.

**Comment by ConverJens** [comment]
@chongkong Pushed fixed for comments. Resolve if your happy.

File: tfx/proto/infra_validator.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (chongkong) [reply tfx/proto/infra_validator.proto:128]: Maybe you forgot to add proto files before commit?

[the remark this answers is not in the document]
Comment (ConverJens) [reply tfx/proto/infra_validator.proto]: I must have missed it. Added it now.

**Comment by ConverJens** [comment]
@chongkong I realised that it is also necessary to be able to load k8s secrets (e.g. Minio credentials) as env vars so I added this functionality to the proto file and kubernetes_runner. Let me know if you think anything should be changed.

File: tfx/proto/infra_validator.proto [hunk 5]
```diff
@@ -119,6 +119,46 @@ message KubernetesConfig {
...
   int32 active_deadline_seconds = 2;
+
+  // Optional.
+  // Additional Kubernetes pod overrides to apply to the serving pod
+  PodManifestOverrides serving_pod_manifest_overrides = 3;
+}
+
+message PodManifestOverrides {
+  // Optional.
+  // Kubernetes annotations to apply to the pod
```
Comment (chongkong) [inline tfx/proto/infra_validator.proto:130]: nit: For comments we have a rule that every sentence (even single) should end with period, and the number of columns doesn't exceed 80. Would you apply such lint for the CL?

File: tfx/proto/infra_validator.proto [hunk 6]
```diff
@@ -119,6 +119,46 @@ message KubernetesConfig {
...
+  // Optional.
+  // Kubernetes environment variables to apply to the pod
+  map<string, string> environment_variables = 2;
+
+  // Optional.
+  // Mapping from Kubernetes secret to environment variable to apply to the pod
+  repeated PodSecretToEnvironmentMapping pod_secret_to_environment_mapping = 3;
+}
+
+message PodSecretToEnvironmentMapping {
```
Comment (chongkong) [inline tfx/proto/infra_validator.proto:142]: This message seems a bit cumbersome. Sorry for your extra work, but let's mimic the Kubernetes EnvVar:

```proto
message PodOverrides {
  map<string, string> annotations = 1;
  repeated EnvVar env = 2;
}
message EnvVar {
  string name = 1;
  string value = 2;
  EnvVarSource valueFrom = 3;
}
message EnvVarSource {
  SecretKeySelector secretKeyRef = 4;
  reserved 1, 2, 3;
}
message SecretKeySelector {
  string name = 1;
  string key = 2;
}
```

Reference: [V1EnvVar](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1EnvVar.md)

Comment (ConverJens) [reply tfx/proto/infra_validator.proto:142]: I think it's valid point and that your proposal is easier and more intuitive for users. Will fix.

Comment (ConverJens) [reply tfx/proto/infra_validator.proto]: Fixed this.

File: tfx/proto/infra_validator.proto (hunk 5 above)
Comment (ConverJens) [reply tfx/proto/infra_validator.proto:130]: Fixed this.

**Comment by ConverJens** [comment]
@chongkong I've pushed the changes you suggested, but when I test it in my pipeline I get a strange error which I do not get with my own custom IV:
```
INFO:absl:Server pod has not been created.
ERROR:absl:Infra validation (attempt 1/1) failed.
Traceback (most recent call last):
  File "/root/pyenv/lib/python3.7/site-packages/tfx/components/infra_validator/executor.py", line 321, in _ValidateWithRetry
    requests=requests)
  File "/root/pyenv/lib/python3.7/site-packages/tfx/components/infra_validator/executor.py", line 356, in _ValidateOnce
    runner.Start()
  File "/root/pyenv/lib/python3.7/site-packages/tfx/components/infra_validator/model_server_runners/kubernetes_runner.py", line 141, in Start
    body=self._BuildPodManifest())
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api/core_v1_api.py", line 7320, in create_namespaced_pod
    return self.create_namespaced_pod_with_http_info(namespace, body, **kwargs)  # noqa: E501
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api/core_v1_api.py", line 7429, in create_namespaced_pod_with_http_info
    collection_formats=collection_formats)
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 353, in call_api
    _preload_content, _request_timeout, _host)
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 170, in __call_api
    body = self.sanitize_for_serialization(body)
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 242, in sanitize_for_serialization
    for key, val in six.iteritems(obj_dict)}
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 242, in <dictcomp>
    for key, val in six.iteritems(obj_dict)}
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 242, in sanitize_for_serialization
    for key, val in six.iteritems(obj_dict)}
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 242, in <dictcomp>
    for key, val in six.iteritems(obj_dict)}
  File "/root/pyenv/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 238, in sanitize_for_serialization
    for attr, _ in six.iteritems(obj.openapi_types)
AttributeError: 'google.protobuf.pyext._message.ScalarMapContainer' object has no attribute 'openapi_types'
INFO:absl:Model failed infra validation.
```

I've tried running both with and without the new pod overrides but I get the same result so It seems to be something apart from that.

Do you have any idea why I get this?

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py [hunk 7]
```diff
@@ -210,6 +211,30 @@ def _BuildPodManifest(self) -> k8s_client.V1Pod:
...
+
+      if self._serving_pod_manifest_overrides and \
+              self._serving_pod_manifest_overrides.env:
+        additional_env_vars = []
+        for env in self._serving_pod_manifest_overrides.env:
+          if env.value:
+            logging.info(f'name={env.name} value={env.value}')
+            additional_env_vars.append(
+              k8s_client.V1EnvVar(name=env.name, value=env.value))
+          else:
```
Comment (chongkong) [inline tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:223]: nit: Let's guard this branch by `elif env.valueFrom and env.valueFrom.secretKeyRef:`

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py [hunk 8]
```diff
@@ -210,6 +211,30 @@ def _BuildPodManifest(self) -> k8s_client.V1Pod:
           model_path=self._model_path)
       env_vars = [k8s_client.V1EnvVar(name=key, value=value)
                   for key, value in env_vars_dict.items()]
+
+      if self._serving_pod_manifest_overrides and \
+              self._serving_pod_manifest_overrides.env:
+        additional_env_vars = []
```
Comment (chongkong) [inline tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:217]: nit: It seems we can just append to `env_vars` without using this variable?

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py [hunk 9]
```diff
@@ -225,6 +250,7 @@ def _BuildPodManifest(self) -> k8s_client.V1Pod:
     result = k8s_client.V1Pod(
         metadata=k8s_client.V1ObjectMeta(
             generate_name=_MODEL_SERVER_POD_NAME_PREFIX,
+            annotations=self._serving_pod_manifest_overrides.annotations,
```
Comment (chongkong) [inline tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:253]: Regarding your error: maybe we should wrap this value inside `dict(self._serving_pod_manifest_overrides.annotations)` to make it serializable by k8s REST client.

**Review by chongkong — CHANGES_REQUESTED** [review]

File: tfx/components/infra_validator/model_server_runners/kubernetes_runner.py (hunk 9 above)
Comment (ConverJens) [reply tfx/components/infra_validator/model_server_runners/kubernetes_runner.py:253]: That fixed it, great catch!

**Comment by ConverJens** [comment]
@chongkong Fix pushed. I'm happy functionality-wise. IV should now have a much greater on-prem and KubeFlow support. Let me know if you have any final remarks.

**Review by chongkong — APPROVED** [review]
Thanks! Will resolve some nitty details from my side :) We don't use github PR directly to merge (because our source of truth for the code stays inside the corp), so please give us a time.

**Comment by ConverJens** [comment]
> Thanks! Will resolve some nitty details from my side :) We don't use github PR directly to merge (because our source of truth for the code stays inside the corp), so please give us a time.

@chongkong Of course, no worries. Pleasure working with you. Thanks for the help!

**Comment by chongkong** [comment]
Hi @ConverJens the change was just submitted but during the internal review process some message and field name has been changed. Please check the final state. Thanks for the contribution!

This change will be available from tfx 0.30.0.

**Pull request merged**